mlflow/transformers.py

```diff
@@ -0,0 +1,851 @@
+    :return: A ``transformers`` model instance or a dictionary of components
+    """
+
+    import transformers
+
+    model_uri = str(model_uri)
+
+    local_model_path = _download_artifact_from_uri(artifact_uri=model_uri, output_path=dst_path)
+
+    flavor_config = _get_flavor_configuration_from_uri(model_uri, FLAVOR_NAME, _logger)
+
+    if return_type == "pipeline" and flavor_config.get(_PROCESSOR_TYPE_KEY, None):
+        raise MlflowException(
+            "This model has been saved with a processor. Processor objects are "
+            "not compatible with Pipelines. Please load this model by specifying "
+            "the 'return_type'='components'.",
+            error_code=BAD_REQUEST,
+        )
+
+    _add_code_from_conf_to_system_path(local_model_path, flavor_config)
+
+    components = _load_model(local_model_path, flavor_config)
+    if return_type == "pipeline":
+        components.update(**kwargs)
+        return transformers.pipeline(**components)
+    elif return_type == "components":
+        return components
+    else:
+        raise MlflowException(
+            f"The specified return_type mode '{return_type}' is unsupported. "
+            "Please select one of: 'pipeline' or 'components'.",
+            error_code=INVALID_PARAMETER_VALUE,
+        )
+
+
+def _load_model(path: str, flavor_config):
+    """
+    Loads components from a locally serialized ``Pipeline`` object.
+    """
+    import transformers
+
+    local_path = pathlib.Path(path)
+    pipeline_path = local_path.joinpath(
+        flavor_config.get(_PIPELINE_BINARY_KEY, _PIPELINE_BINARY_FILE_NAME)
+    )
+
+    model_instance = getattr(transformers, flavor_config[_PIPELINE_MODEL_TYPE_KEY])
+    conf = {
+        "task": flavor_config[_TASK_KEY],
+        "model": model_instance.from_pretrained(pipeline_path),
+    }
+
+    if flavor_config.get(_PROCESSOR_TYPE_KEY, None):
+        conf[_PROCESSOR_KEY] = _load_component(
+            local_path, _PROCESSOR_KEY, flavor_config[_PROCESSOR_TYPE_KEY]
+        )
+
+    for component_key in flavor_config[_COMPONENTS_BINARY_KEY]:
+        component_type_key = f"{component_key}_type"
+        component_type = flavor_config[component_type_key]
+        conf[component_key] = _load_component(local_path, component_key, component_type)
+    return conf
+
+
+def _fetch_model_card(model_or_pipeline):
+    """
+    Attempts to retrieve the model card for the specified model architecture iff the
+    `huggingface_hub` library is installed. If a card cannot be found in the registry or
+    the library is not installed, returns None.
+    """
+    with contextlib.suppress(ImportError):
+        from transformers import Pipeline
+
+        hub = importlib.import_module("huggingface_hub")
+        model = (
+            model_or_pipeline.model
+            if isinstance(model_or_pipeline, Pipeline)
+            else model_or_pipeline[_MODEL_KEY]
+        )
+        return hub.ModelCard.load(model.name_or_path)
```

Comment:
Looks like `hub.ModelCard` isn't defined in version 0.8.1:

```
from transformers import AutoTokenizer, AutoModelWithLMHead, pipeline

import torch

tokenizer = AutoTokenizer.from_pretrained('microsoft/DialoGPT-small')
model = AutoModelWithLMHead.from_pretrained('satvikag/chatbot')

import mlflow.transformers

mlflow.transformers.log_model(
   pipeline(task="conversational", model=model, tokenizer=tokenizer),
   artifact_path="model",
)
```

```
  File "chat.py", line 35, in <module>
    mlflow.transformers.log_model(
  File "/Users/corey.zumar/mlflow/mlflow/transformers.py", line 467, in log_model
    return Model.log(
  File "/Users/corey.zumar/mlflow/mlflow/models/model.py", line 547, in log
    flavor.save_model(path=local_path, mlflow_model=mlflow_model, **kwargs)
  File "/Users/corey.zumar/mlflow/mlflow/transformers.py", line 295, in save_model
    card_data = model_card if model_card is not None else _fetch_model_card(transformers_model)
  File "/Users/corey.zumar/mlflow/mlflow/transformers.py", line 609, in _fetch_model_card
    return hub.ModelCard.load(model.name_or_path)
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/huggingface_hub/__init__.py", line 105, in __getattr__
    raise AttributeError(f"No {package_name} attribute {name}")
AttributeError: No huggingface_hub attribute ModelCard
```

Can we do a hasattr here and print a warning if:

- HuggingFace Hub is not importable (maybe contextlib.suppresss handles this?)
- HuggingFace Hub is too old and doesn't define hub.ModelCard

Reply:
In our docstring, can we also indicate that `huggingface_hub >= {min version that defines ModelCard}` is required and that `transformers >= 4.25.1` is required? Can we use the same docstring notice syntax that we use for autologging? E.g.

<img width="1337" alt="Screen Shot 2023-03-27 at 9 58 27 PM" src="https://user-images.githubusercontent.com/39497902/228132738-5898eee6-bd38-4792-a900-c4a509f69be1.png">

It would be great to autogenerate this from ml-package-versions.yml, which we do for autologging. Feel free to track this as a follow-up ticket to complete before we release :)